draft-ietf-quic-tls.md

```diff
@@ -1059,6 +1059,27 @@ could either be absent or it could indicate that an address validation token is
 not present.
 
 
+### Stateless Address Validation
+
+A server can use the cookie extension to store all state necessary to continue
+the connection.  This allows a server to avoid committing state for clients that
+have unvalidated source addresses.
+
+For instance, a server could use a statically-configured key to encrypt the
+information that it requires and include that information in the cookie.  In
+addition to address validation information, a server that uses encryption also
+needs to be able recover the hash of the ClientHello and its length, plus any
+information it needs in order to reconstruct the HelloRetryRequest.
+
+A client proceeds as though the server were stateful.  It constructs a second
+ClientHello and sends it following the initial ClientHello.  Note that this
+means a server that doesn't maintain state between the intial and second
+ClientHello will receive the second ClientHello in STREAM frames with a non-zero
+offset.  The resulting gap in the received stream data needs to be ignored and
```

Comment:
I can understand how this could be made to work, but I worry a bit about the state of the crypto stream being different between the two endpoints. The client believes it has send 2 ClientHellos as bytes [0,N] on stream 0. If the server is required to process pre-handshake stream frames even with non-0 offsets, then it seems like it might be possible for an active attacker to rewrite the second ClientHello to start at, say, offset 2N. This would mean that any post-handshake messages from the client to the server would be ignored as duplicates.

I presume the solution to this is for the server to include in the data it sends to the client (and the client sends back to the server) the byte offset that the follow-up ClientHello is expected to start at, and to validate that when it process the follow-up ClientHello.

Does that sound right?